Overhaul cursor loading

pchote opened the pull request:
This PR reworks the way we load ~and render~ cursors, ~adding support for 24/32bit artwork and~ improving cross-platform compatibility for hardware cursors.

In particular, this fixes a relatively recent macOS issue (possibly introduced with 10.15) where relative offsets between frames in an animation seem to be ignored, making them jump around incorrectly. If we are lucky, this may also finally solve #10172. See commit messages and code comments for more details.

~Depends on #16885.~

dragunoff commented:
Just tested this. The cursors render but I am seeing some RGBA/BRGA weirdness with the select cursor - it is using the same sequence as the default cursor but it seems to be swapping the blue/red channels when nothing is selected:
![rgba-cursor-weirdness](https://user-images.githubusercontent.com/1355810/71562841-b7776800-2a8e-11ea-9a61-8d0cedea5cba.gif)

dragunoff commented:
Also notice how the cursor is skewed and distorted in the screen capture above. This only affects hardware cursors, the correct cursor size can be seen in the gif below. It renders fine on screen so this is could be due to some weird interaction with the screen capturing software.

The gif below also illustrates that the setting to increase the command cursor size doesn't work well when the default cursor is the same as the move cursor (but that seems like a separate issue):
![17i6xa3LC9](https://user-images.githubusercontent.com/1355810/71563063-5d78a180-2a92-11ea-8345-539417f1e234.gif)

pchote commented:
The mouse cursor issues that this fixes on macOS are rather annoying, so I will separate the platform workarounds from the RGBA support (which can go back into #16885, inverting the PR dependency) so we can take this for the hotfix.

pchote commented:
Updated. This now includes only the `HardwareCursor` fixes, and picks cleanly to prep.

Mailaender reviewed (dismissed):
I spot no regressions on Linux.

pchote commented on OpenRA.Game/Graphics/HardwareCursor.cs:
```diff
@@ -38,74 +37,98 @@ public HardwareCursor(CursorProvider cursorProvider)
+					frameSprites[i] = sheetBuilder.Add(frame.Data, frame.Size, 0, frame.Offset);
+
+					// CreateHardwareCursor expects data in RGBA
+					frameData[i] = FrameToRGBA(frame, palette);
```
It would simplify life here and in #17517 if we move this conversion inside CreateCursor.

teinarss commented on OpenRA.Game/Graphics/HardwareCursor.cs:
```diff
@@ -38,74 +37,98 @@ public HardwareCursor(CursorProvider cursorProvider)
 
-				hardwareCursors.Add(kv.Key, hc);
+				var paddedSize = 8 * new int2((sequenceBounds.Width + 7) / 8, (sequenceBounds.Height + 7) / 8);
+				for (var i = 0; i < frames.Length; i++)
```
Im guessing that we need to do this in two stages because how `paddedSize` is calculated, right?

pchote replied:
Correct. I'll move the `frameSprites` and `frameData` handling into this loop to make the separation of concerns clearer.

teinarss commented on OpenRA.Game/Graphics/HardwareCursor.cs:
```diff
@@ -38,74 +37,98 @@ public HardwareCursor(CursorProvider cursorProvider)
+				// - the hotspot is inside the frame bounds (enforced by SDL)
+				// - all frames within a sequence have the same size (needed for macOS 10.15)
+				// - the frame size is a multiple of 8 (needed for Windows)
+				Rectangle sequenceBounds = Rectangle.FromLTRB(0, 0, 1, 1);
```
var?

pchote committed:
Rework HardwareCursor sprite padding.

All frames in a sequence now use the same bounds
and hotspot, and have a size that is a multiple of 8.

pchote commented:
Updated.

abcdefg30 reviewed (approved):
#10172 seems to no longer happen when starting D2k, but still happens around every second time starting TS.

abcdefg30 commented:
Still merging since fixing #10172 was not the primary goal here.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/1a50c672c46100f12ec9c584934c4a38666c7049...d45bac834e17f4ce7a39786e5b510130280bd835)

https://github.com/OpenRA/OpenRA/commit/1b3d8dfc8db5190bad850c84eaa31fa07c98d665
